gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/catalog/GfxdSystemProcedures.java

```diff
@@ -2466,11 +2484,17 @@ public static void USE_SNAPSHOT_TXID(String txId) throws SQLException {
     TXStateInterface state = tc.getTransactionManager().getHostedTXState(txId1);
 
     if (state == null) {
+      if (GemFireXDUtils.TraceExecution) {
+      SanityManager.DEBUG_PRINT(GfxdConstants.TRACE_EXECUTION,
+          "in procedure USE_SNAPSHOT_TXID() creating a txState for conn " + tc.getConnectionID() + " tc id" + tc.getTransactionIdString()
+              + " txId  " +txId);
+      }
       // if state is null then create txstate and use
       state =  tc.getTransactionManager().getOrCreateHostedTXState(txId1,
           com.gemstone.gemfire.internal.cache.locks.LockingPolicy.SNAPSHOT, true);
     }
     tc.getTransactionManager().masqueradeAs(state);
+    TXManagerImpl.snapshotTxState.set(state);
```

Comment:
Why should explicit calls to TXManagerImpl.snapshot be done? It should be taken care of by masqueradeAs if the state is a snapshot one. We need to cleanup this handling and combine into a single TXState tracking in thread-local.